### ISSUE TITLE
useFocusWithin ignores events from portals

### ISSUE THREAD

**Issue by 42shadow42** [adobe/react-spectrum#2733]
# 🐛 Bug Report 

When using portals to place elements inside the focus boundary, focus can enter/exit the focus boundary without detection.

## 🤔 Expected Behavior

One of two behaviors are expected, the elements from the portal should be considered inside or outside of the boundary and handled accordingly. To Clarify

Option 1 (Preferred) Elements are considered inside the focus boundary
When elements from the portal receive focus from outside the focus boundary the onFocusWithin event should be fired.
When elements from the portal blur without moving to another element inside the focus boundary the onBlurWithin event should be fired.

Option 2 Elements are considered outside the focus boundary
When elements from the portal receive focus from outside the focus boundary the onFocusWithin event should not be fired
When elements from the portal are focused while losing focus within the focus boundary the onBlurWithin event should be fired

## 😯 Current Behavior
When elements from the portal receive focus from outside the focus boundary the onFocusWithin event is not fired (bug for option 1)
When elements from the origin receive focus from outside the focus boundary the onFocusWithin event is fired (actual behavior)
When elements from the origin receive focus from elements from the portal the onFocusWithin event is fired (bug for option 1)

When elements from the portal lose focus without focusing an element in the focus boundary the onBlurWithin event is not fired (bug for option 1)
When elements from the origin lose focus without focusing an element in the focus boundary the onBlurWithin event is fired (actual behavior)
When elements from the origin lose focus with focus moving to an element from the portal the onBlurWithin event is not fired (bug for option 2)

## 🔦 Context

I was trying to create accessibility shortcuts that move focus to important areas of a page with many focusable elements. The portal allowed a way of collecting all the shortcuts together into one place on the dom so they are tab ordered before the content without the use of tabindex (anything different from 0 or -1 is considered bad practice imo) the shortcuts are provided through a portal so they are defined in the context they refer to allowing easy access to the .focus() function on the desired landing element.

## 💻 Code Sample

https://codesandbox.io/s/snowy-sunset-eh0j8

## 🌍 Your Environment

| Software         | Version(s) |
| ------------------ | ---------- |
| react-spectrum     |  3.7.0
| react                       | 17.0.2
| Browser                  | Chrome
| Operating System | Mac

**Comment by LFDanLu** [comment]
Is the codesandbox you provided a roughly accurate setup of what you have in your app? Is there a particular reason why the first button + the wrapping div aren't within the Portal instead?

**Comment by 42shadow42** [comment]
No, it's something I had to simplify in order to make it viable for a reproducible sandbox. The reason the first button isn't within the portal is to demonstrate the difference in behavior. The actual use case they are all inside.

**Comment by LFDanLu** [comment]
Perhaps this is something with React since the outer div isn't an ancestor to the portalled buttons in the React tree. Replacing the useFocusWithin event handlers with DOM event handlers seems to make it work.

**Comment by 42shadow42** [comment]
> Perhaps this is something with React since the outer div isn't an ancestor to the portalled buttons in the React tree. Replacing the useFocusWithin event handlers with DOM event handlers seems to make it work.

I wondered about that, react uses virtual events that bubble in a different hierarchy I think. I noticed when I put in logs that the events necessary didn't even fire. I also found that using the dom events helped but was only getting the blur event not the focus event. I think this is because the focus dom event doesn't bubble.

**Comment by LFDanLu** [comment]
Right, you'll have to use `focusin` instead of `focus` for bubbling. I'll note this one for issue grooming.

**Comment by LFDanLu** [comment]
Quick question: could you instead have the focusWithin event handlers on div that wraps the buttons within the Portal instead?

**Comment by 42shadow42** [comment]
Maybe? I'd have to try that I hadn't thought of it.

**Comment by LFDanLu** [comment]
Closing as stale, but let us know if the above doesn't work for you and we can reopen
